sbpy/thermal/core.py

```diff
@@ -5,66 +5,274 @@
-        >>> eph = Ephem.from_horizons('2015 HW', location='568', epochs=epoch) # doctest: +REMOTE_DATA
-        >>> phys = PhysProp('diam'=0.3*u.km, 'pv'=0.3) # doctest: +SKIP
-        >>> lam = np.arange(1, 20, 5)*u.micron # doctest: +SKIP
-        >>> flux = STM.flux(phys, eph, lam) # doctest: +SKIP
-
-        not yet implemented
+        lon : float
+            Longitude in radiance
+        lat : float
+            Latitude in fradiance
+        m : numpy array of shape (3, 3)
+            Transformation matrix to convert a vector in the frame to perform
+            integration to body-fixed frame.  This matrix can be calculated
+            with private method `_transfer_to_bodyframe`.  The integration
+            is performed in a frame where the sub-observer point is defined at
+            lon = 0, lat = 0.
+        unit : str or astropy.units.Unit
+            Unit of the integral function.
+        wave_freq : u.Quantity
+            Wavelength or frequency of calculation
 
+        Returns
+        -------
+        float : Integral function to calculate total flux density.
         """
+        _, lon1, lat1 = xyz2sph(
+            m.dot(sph2xyz([np.rad2deg(lon), np.rad2deg(lat)]))
+            )
+        lon1 *= u.deg
+        lat1 *= u.deg
+        T = self.T(lon1, lat1)
+        if np.isclose(T, 0 * u.K):
+            return 0.
+        else:
+            # the integral term needs to include a correction for latitude
+            # with cos(lat), and a Lambertian emission term cos(lat) + cos(lon)
+            coslat = np.cos(lat)
+            coslon = np.cos(lon)
+            f = BlackBody(T)(wave_freq) * coslat * coslat * coslon
+            return f.to_value(unit, u.spectral_density(wave_freq))
 
-    def fit(self, eph):
-        """Fit thermal model to observations stored in `sbpy.data.Ephem` instance
+    @staticmethod
+    @u.quantity_input(sublon=u.deg, sublat=u.deg)
+    def _transfer_to_bodyframe(sublon, sublat):
+        """Calculate transformation matrix.
+
+        The numerical integration to calculate total flux density is performed
+        in a reference frame where the sub-observer point is at
+        (lon, lat) = (0, 0).  This matrix supports the transformation from
+        this frame to the body-fixed frame to facilitate the calculation of
+        surface temperature.
+        """
+        coslat = np.cos(sublat).value
+        if abs(coslat) < np.finfo(type(coslat)).resolution:
+            if sublat.value > 0:
+                m = np.array([[0, 0, -1], [0, 1, 0], [1, 0, 0]])
+            else:
+                m = np.array([[0, 0, 1], [0, 1, 0], [-1, 0, 0]])
+        else:
+            m = twovec([sublon.to_value('deg'), sublat.to_value('deg')], 0,
+                       [0, 90], 2).T
+        return m
+
+    @u.quantity_input(wave_freq=u.m, delta=u.m, lon=u.deg, lat=u.deg,
+                      equivalencies=u.spectral())
+    def fluxd(self, wave_freq, delta, sublon, sublat, unit='W m-2 um-1',
```

Comment:
I like the idea of using sublon and sublat, but can we also have a way to use obliquity and phase angle?

Reply:
This method calculates the total flux density by integrating over the whole visible part of the surface of an object, provided that the temperature distribution is available by `.T(lon, lat)`.  So as long as a mechanism is provided to convert (obliquity, phase) or any other combination to (sublon, sublat) in a subclass or inherited `.fluxd` method, we can use it in any circumstance.  I think this will also work for an irregular shape model, but haven't thought about it thoroughly or tested it yet.